client/executor/wasmtime/src/runtime.rs

```diff
@@ -51,41 +76,62 @@ impl WasmModule for WasmtimeRuntime {
 
 		// Scan all imports, find the matching host functions, and create stubs that adapt arguments
 		// and results.
+		//
+		// NOTE: Attentive reader may notice that this could've been moved in `WasmModule` creation.
+		//       However, I am not sure if that's a good idea since it would be pushing our luck further
+		//       by assuming that `Store` not only `Send` but also `Sync`.
 		let imports = resolve_imports(
 			&store,
-			self.module_wrapper.module(),
+			&self.module,
 			&self.host_functions,
-			self.heap_pages,
-			self.allow_missing_func_imports,
+			self.config.heap_pages,
+			self.config.allow_missing_func_imports,
 		)?;
 
-		let instance_wrapper =
-			InstanceWrapper::new(&store, &self.module_wrapper, &imports, self.heap_pages)?;
-		let heap_base = instance_wrapper.extract_heap_base()?;
-		let globals_snapshot = GlobalsSnapshot::take(&instance_wrapper)?;
-
-		Ok(Box::new(WasmtimeInstance {
-			store,
-			instance_wrapper: Rc::new(instance_wrapper),
-			module_wrapper: Arc::clone(&self.module_wrapper),
-			imports,
-			globals_snapshot,
-			heap_pages: self.heap_pages,
-			heap_base,
-		}))
+		let strategy = if self.config.semantics.fast_instance_reuse {
+			let instance_wrapper =
+				InstanceWrapper::new(&store, &self.module, &imports, self.config.heap_pages)?;
+			let heap_base = instance_wrapper.extract_heap_base()?;
+
+			let mut_globals = self
+				.mutable_globals
+				.as_ref()
+				.expect("mutable_globals are Some when fast_instance_reuse is set; qed");
+
+			// This function panics if the instance was created from a runtime blob different from which
+			// the mutable globals were collected. Here, it is easy to see that there is only a single
+			// runtime blob and thus it's the same that was used for both creating the instance and
+			// collecting the mutable globals.
+			let globals_snapshot = GlobalsSnapshot::take(mut_globals, &instance_wrapper);
+
+			let data_segments_snapshot = self
+				.data_segments_snapshot
+				.clone()
+				.expect("data_segments_shanpshot is Some when fast_instance_reuse is set; qed");
+
+			Strategy::FastInstanceReuse {
+				instance_wrapper: Rc::new(instance_wrapper),
+				globals_snapshot,
+				data_segments_snapshot,
+				heap_base,
+			}
+		} else {
+			Strategy::RecreateInstance(InstanceCreator {
```

Comment:
This could be moved above the if statement and also reused in the upper branch.

Reply:
I am not sure if I follow: how? Those use distinct data, so what would be reused here?

Reply:
Not sure anymore what I wanted here xD